components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissions.kt

```diff
@@ -7,31 +7,78 @@ package mozilla.components.feature.sitepermissions
 import android.os.Parcel
 import android.os.Parcelable
 import mozilla.components.feature.sitepermissions.SitePermissions.Status.NO_DECISION
+import mozilla.components.feature.sitepermissions.SitePermissionsStorage.Permission
+import mozilla.components.feature.sitepermissions.SitePermissionsStorage.Permission.BLUETOOTH
+import mozilla.components.feature.sitepermissions.SitePermissionsStorage.Permission.CAMERA
+import mozilla.components.feature.sitepermissions.SitePermissionsStorage.Permission.MICROPHONE
+import mozilla.components.feature.sitepermissions.SitePermissionsStorage.Permission.NOTIFICATION
+import mozilla.components.feature.sitepermissions.SitePermissionsStorage.Permission.LOCATION
+import mozilla.components.feature.sitepermissions.SitePermissionsStorage.Permission.LOCAL_STORAGE
 import mozilla.components.feature.sitepermissions.db.StatusConverter
 
 /**
  * A site permissions and its state.
  */
 data class SitePermissions(
     val origin: String,
-    val location: Status = NO_DECISION,
-    val notification: Status = NO_DECISION,
-    val microphone: Status = NO_DECISION,
-    val camera: Status = NO_DECISION,
-    val bluetooth: Status = NO_DECISION,
-    val localStorage: Status = NO_DECISION,
+    val permissions: Map<Permission, Status> = emptyMap(),
```

Comment:
I was thinking that we could find **shorter path** that lead of to **less code to maintain**,  **less code to modify** wand the same end result.

What about if instead of deprecating the constructor we just add a `get` operator  that relies on the existing properties? 

```kotlin
    /**
     * Gets the current status for a [Permission] type
     */
    operator fun get(permissionType: Permission): Status {
        return when(permissionType){
            MICROPHONE -> microphone
            BLUETOOTH -> bluetooth
            CAMERA -> camera
            LOCAL_STORAGE -> localStorage
            NOTIFICATION -> notification
            LOCATION -> location
        }
    }
```
This way we can do:

```kotlin
    val sitePermission = SitePermissions(
            origin = "mozilla.dev",
            savedAt = 0
    )
    sitePermission[NOTIFICATION]
```

Without having to maintain our own `equals` and `hashCode` functions as they come for free form the `data class` and at the same time we minimize the changes on the consumers end.

![image](https://user-images.githubusercontent.com/773158/68045102-231b9f80-fcaf-11e9-98eb-cd6d25975395.png)


@calvin-li what do you think?

Reply:
So, I've never used data classes before and didn't realize they had Equals() built-in. Looks like we could just use the default implementation then? I can try it later when I'm at home. 

We should already have the `site[NOTIFICATION]` behavior, if you're concerned about having to change usages in the rest of the code we can just not deprecate the old constructor?

Reply:
> So, I've never used data classes before and didn't realize they had Equals() built-in. Looks like we could just use the default implementation then? I can try it later when I'm at home.

Yes, [these are the nice things about data classes](https://kotlinlang.org/docs/reference/data-classes.html) :)

> We should already have the site[NOTIFICATION] behavior, if you're concerned about having to change usages in the rest of the code we can just not deprecate the old constructor?

Yes, I think we could just leave the `SitePermissions` as it was before and add the `get` operator.

Reply:
@NotWoods can clairfy, but I one of the reasons the issue was opened in the first place was the difficulty serializing the class. Adding `get` like this wouldn't help. This also means we would still have to update the `SitePermissions` class every time a new `Permission` type is added, which IMO is another good reason to use the enums.

Also, how would the current changes (bba4cce) make the api easier to use? Does it offer anything over leaving the class as-is, and is it enough to justify going through the testing/review process? Maybe we shouldn't be doing anything here at all.

Finally, in the future it would be nice to have these kinds of discussions on the first draft of the pull request, before all the work for testing, updating usages elsewhere, etc are already done